### DISCUSSION TITLE
v3 - Pagination with nested Livewire components

### DISCUSSION THREAD

**Discussion by scottjo** [livewire/livewire#7683] [1. Help]
Good morning folks - newbie here so please be gentle.

 I am in the process of writing an e-commerce page where I have a paginated list of products. So within my catalogue page wrapper Livewire component I am querying my product table with pagination with a page size of 10 items. 

I build up the query applying different filters etc. in a method on the wrapper - the whole component has the WithPagination trait. In my blade component I iterate over the products and render a blade template for each product.

This all works a treat. No issues. 

However, if I replace that product card with a Livewire component OR add a Livewire component to the product card, the pagination goes belly up.  I get the first page no problem, As soon as I go to the next page, in the console,  I get the uncaught component not found and a large LW id. I also only get 8 items rendered before this crashes and burns.

I'm giving the livewire component that I am adding to the product card a wire:key of the product_id -

`<livewire:catalogue.add-to-favourites :product_id="$product->id" :key="$product->id" />`

and the template that is being rendered is 

```
<button type="button" title="Save to favourites" wire:click="addToFavourite">

    <x-icon name="o-heart"
            @class( ["cursor-pointer",
                    "fill-red-500" => $in_wish_list]) />

</button>
```

As you can see it is extremely simple.  if I inspect the node in the Dom should I see that wire:key?  I tried adding the wire:key directive to the button declaration but that made no difference.

What am I doing wrong - surely it can't be this difficult...

Jon

**Comment by joshhanley** [comment]
@scottjo can you show your code for the loop and nested livewire component all together including where you tried keys?

**Comment by scottjo** [comment]
Hi Josh

So in my 'outer' Livewire component I have

```
                    <div class="flex flex-wrap p-0 my-0 mr-0 align-baseline border-0 text-neutral-800 gap-3">
                        @foreach( $products as $product )
                            @include('partials.catalogue.small-product-card', ['product' => $product])
                        @endforeach

                        @include('partials.catalogue.small-valuation-card')
                    </div>

                    {{ $products->links() }}
```

Previously this was a Livewire component but had similar issues so reverted to a plain and simple blade file. 

Then within that individual blade file I have 

```
        <livewire:catalogue.add-to-favourites 
                :product_id="$product->id" 
                :key="$product->id" />
```

And the blade file for the add-to-favourites is what I pasted initially. Note as the loop in the parent lw component is just iterating a collection and rendering a non Livewire blade file I've not given it a key etc. 

So my hierarchy is

Livewire wrapper
      -> n number of blade templates
                 -> each blade template has a Livewire component

My paginated products are being generated in the Livewire wrapper

```
    public function render()
    {
        return view('livewire.catalogue.catalogue', ['products' => $this->getProducts()]);
    }

```

The getProducts function on the Livewire wrapper is basically an encapsulated Eloquent builder that, at the end, I am saying

```
return $query->paginate(10);
```

**Reply by joshhanley** [reply]
@scottjo right yep, that makes sense. So the issue will be a morphing issue and probably needing keys in slightly different places.

I'm not exactly clear on what your `small-product-card` contains, but this is how I would structure it (without the blade component)

```blade
<div class="flex flex-wrap p-0 my-0 mr-0 align-baseline border-0 text-neutral-800 gap-3">
    @foreach( $products as $product )
        <div wire:key="{{ $product->id }}">
            @include('partials.catalogue.small-product-card', ['product' => $product])
        </div>
    @endforeach

    <div wire:key="small-valuation-card">
        @include('partials.catalogue.small-valuation-card')
    </div>
</div>
```

See how for elements in a loop, I have one root element which has the key on it? And I also have added one for the element outside the loop, but that's contained within the same parent element as the loop.

Then the other thing I would do, is on your nested component, just change the key a little so it's different to the keys in the loop:

```blade
<livewire:catalogue.add-to-favourites 
                :product_id="$product->id" 
                wire:key="add-favourite-{{ $product->id }}" />
```

**Reply by scottjo** [reply]
@joshhanley - thanks so much - I think the key (no pun intended) is to wrap the whole include of the product card with the key'd div. That seems to have fixed it. the child component on the product card is also working as expected. Thanks so much for your help. Much appreciated.

**Reply by joshhanley** [reply]
@scottjo sweet! No worries! 🙂
